From a425d9ceb30bc9d8bd692a643494ecbf52fe70f0 Mon Sep 17 00:00:00 2001
From: Gal Savion <gsavion@maxlinear.com>
Date: Fri, 3 May 2024 18:20:43 +0300
Subject: [PATCH] WLANRTSYS-80477 implement Wifi.AccessPoint.vap5g0priv
 cleanStation()

Implement the option to trigger hostapd/driver to clear an inactive STA
from the STA databases at all layers in the AP (hostapd, driver, FW),
without sending any frames (data, deauth or disassoc) to the STA.

This PR resolves an issue at hostapd, that is still sending DISASSOC
frame (after quiet DEAUTH) or DEAUTH frame (after quiet DISASSOC) to
the client after some inactivity timeout, where the original command
has tx=0 parameter.
---
 src/ap/ctrl_iface_ap.c | 44 +++++++++++++++++++++++++-----------------
 1 file changed, 26 insertions(+), 18 deletions(-)

diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index 8f93727ee..df2e6f0d7 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -644,20 +644,24 @@ int hostapd_ctrl_iface_deauthenticate(struct hostapd_data *hapd,
 	}
 #endif /* CONFIG_P2P_MANAGER */
 
-	if (os_strstr(txtaddr, " tx=0"))
+	sta = ap_get_sta(hapd, addr);
+
+	if (os_strstr(txtaddr, " tx=0")) {
 		hostapd_drv_sta_remove(hapd, addr);
+		if (sta)
+			ap_free_sta(hapd, sta);
 #ifdef CONFIG_VENDOR_MXL
-	else if (is_broadcast_ether_addr(addr) &&
-		(hapd->wpa_auth) && (hapd->conf->ieee80211w != NO_MGMT_FRAME_PROTECTION))
+	} else if (is_broadcast_ether_addr(addr) &&
+		(hapd->wpa_auth) && (hapd->conf->ieee80211w != NO_MGMT_FRAME_PROTECTION)) {
 		mxl_hostapd_send_protected_deauth(hapd, reason);
 #endif /* CONFIG_VENDOR_MXL */
-	else
+	} else {
 		hostapd_drv_sta_deauth(hapd, addr, reason);
-	sta = ap_get_sta(hapd, addr);
-	if (sta)
-		ap_sta_deauthenticate(hapd, sta, reason);
-	else if (addr[0] == 0xff)
-		hostapd_free_stas(hapd);
+		if (sta)
+			ap_sta_deauthenticate(hapd, sta, reason);
+		else if (addr[0] == 0xff)
+			hostapd_free_stas(hapd);
+	}
 
 	return 0;
 }
@@ -718,20 +722,24 @@ int hostapd_ctrl_iface_disassociate(struct hostapd_data *hapd,
 	}
 #endif /* CONFIG_P2P_MANAGER */
 
-	if (os_strstr(txtaddr, " tx=0"))
+	sta = ap_get_sta(hapd, addr);
+
+	if (os_strstr(txtaddr, " tx=0")) {
 		hostapd_drv_sta_remove(hapd, addr);
+		if (sta)
+			ap_free_sta(hapd, sta);
 #ifdef CONFIG_VENDOR_MXL
-	else if (is_broadcast_ether_addr(addr) &&
-		(hapd->wpa_auth) && (hapd->conf->ieee80211w != NO_MGMT_FRAME_PROTECTION))
+	} else if (is_broadcast_ether_addr(addr) &&
+		(hapd->wpa_auth) && (hapd->conf->ieee80211w != NO_MGMT_FRAME_PROTECTION)) {
 		mxl_hostapd_send_protected_disassoc(hapd, reason);
 #endif /* CONFIG_VENDOR_MXL */
-	else
+	} else {
 		hostapd_drv_sta_disassoc(hapd, addr, reason);
-	sta = ap_get_sta(hapd, addr);
-	if (sta)
-		ap_sta_disassociate(hapd, sta, reason);
-	else if (addr[0] == 0xff)
-		hostapd_free_stas(hapd);
+		if (sta)
+			ap_sta_disassociate(hapd, sta, reason);
+		else if (addr[0] == 0xff)
+			hostapd_free_stas(hapd);
+	}
 
 	return 0;
 }
-- 
2.43.0

